samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -188,98 +183,82 @@ private void validateConfig() {
   }
 
   /**
-   * This function traverses the {@link OperatorSpec} graph to find and update mappings for all Joins reachable
-   * from this input {@link StreamEdge}.
-   * @param operatorSpec the {@link OperatorSpec} to traverse
-   * @param sourceStreamEdge source {@link StreamEdge}
-   * @param joinSpecToStreamEdges mapping from join spec to its source {@link StreamEdge}s
-   * @param streamEdgeToJoinSpecs mapping from source {@link StreamEdge} to the join specs that consumes it
-   * @param joinQ queue that contains joinSpecs where at least one of the input stream edge partitions is known.
+   * Creates a {@link JoinedStreamsGroup} whose Id is {@code groupId}, and {@link StreamEdge}s
+   * correspond to the provided {@code inputOperatorSpecs}.
    */
-  private static void findReachableJoins(OperatorSpec operatorSpec, StreamEdge sourceStreamEdge,
-      Multimap<JoinOperatorSpec, StreamEdge> joinSpecToStreamEdges,
-      Multimap<StreamEdge, JoinOperatorSpec> streamEdgeToJoinSpecs,
-      Queue<JoinOperatorSpec> joinQ, Set<JoinOperatorSpec> visited) {
-
-    if (operatorSpec instanceof JoinOperatorSpec) {
-      JoinOperatorSpec joinOperatorSpec = (JoinOperatorSpec) operatorSpec;
-      joinSpecToStreamEdges.put(joinOperatorSpec, sourceStreamEdge);
-      streamEdgeToJoinSpecs.put(sourceStreamEdge, joinOperatorSpec);
-
-      if (!visited.contains(joinOperatorSpec) && sourceStreamEdge.getPartitionCount() > 0) {
-        // put the joins with known input partitions into the queue and mark as visited
-        joinQ.add(joinOperatorSpec);
-        visited.add(joinOperatorSpec);
+  private JoinedStreamsGroup getJoinedStreamsGroup(String groupId, Iterable<InputOperatorSpec> inputOperatorSpecs,
+      JobGraph jobGraph) {
+
+    int countStreamEdgeWithSetPartitions = 0;
+    Set<StreamEdge> streamEdges = new HashSet<>();
+
+    for (InputOperatorSpec inputOpSpec : inputOperatorSpecs) {
+      StreamEdge streamEdge = jobGraph.getOrCreateStreamEdge(getStreamSpec(inputOpSpec.getStreamId(), streamConfig));
+      if (streamEdge.getPartitionCount() != StreamEdge.PARTITIONS_UNKNOWN) {
+        ++countStreamEdgeWithSetPartitions;
       }
+      streamEdges.add(streamEdge);
     }
 
-    Collection<OperatorSpec> registeredOperatorSpecs = operatorSpec.getRegisteredOperatorSpecs();
-    for (OperatorSpec registeredOpSpec : registeredOperatorSpecs) {
-      findReachableJoins(registeredOpSpec, sourceStreamEdge, joinSpecToStreamEdges, streamEdgeToJoinSpecs, joinQ,
-          visited);
+    // Determine category of stream group based on stream partition counts.
+    JoinedStreamsGroupCategory category;
+    if (countStreamEdgeWithSetPartitions == 0) {
+      category = JoinedStreamsGroupCategory.NO_PARTITION_COUNT_SET;
+    } else if (countStreamEdgeWithSetPartitions == streamEdges.size()) {
+      category = JoinedStreamsGroupCategory.ALL_PARTITION_COUNT_SET;
+    } else {
+      category = JoinedStreamsGroupCategory.SOME_PARTITION_COUNT_SET;
     }
+
```

Comment:
The notion of a "category" looks like a detail of the `JoinedStreamsGroup`, which can be inferred from the other params. What do you think about moving the logic that determines "category" to the constructor of `JoinedStreamsGroup` ?

A nice property is that we could avoid inconsistent object states. For eg: with the current constructor: `JoinedStreamsGroup(groupId, streamEdges, category)`, one could create an instance of `JoinedStreamsGroup` such that its "streamEdges" and "category" contradict each other

Reply:
I did consider this and I would have really liked to do so. However,

1. Deciding the category in `JoinedStreamsGroup` requires one more (redundant) iteration over the `StreamEdge`s.

2. More importantly, there is no way to avoid inconsistent object states in `JoinedStreamsGroup` because `StreamEdge`s are mutable. In fact, we already throw every `JoinedStreamsGroup` with  intermediate streams into this inconsistent state once we start setting partition counts. This made me less inclined to incur the overhead of the extra iteration in #1.

Together, these 2 points made me prefer to keep `JoinedStreamsGroup` as a passive data object that contains zero logic. I thought this would help set the readers` expectations that this object is just a way of organizing data.